File: arrays_matrices/prob1365.py
# how many numbers smaller than current numbers

class Solution(object):
    def smallerNumbersThanCurrent(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        histo = [0] * 101
        for n in nums:
            histo[n] += 1
        csum = 0 # cumulative sum
        for i in range(101):
            prev = histo[i]
            histo[i] = csum
            csum += prev
        
        return [ histo[n] for n in nums]  

File: hashtable/prob1365.kt
// 1365 How many numbers are smaller than the current number

class Solution {
    fun smallerNumbersThanCurrent(nums: IntArray): IntArray {

        val n = nums.size
        val ans = IntArray(n)

        val hm = mutableMapOf<Int, MutableSet<Int>>()

        for (i in 0 until n) {
            var v = nums[i]
            if (v in hm) {
                hm[v]?.add(i)
            } else {
                hm[v] = mutableSetOf<Int>(i)
            }
        }

        var pairs = hm.toList().sortedBy {(first, second) -> first}

        var psum = 0
        val m = pairs.size

        for (i in 1 until m) {
            psum += pairs[i-1].second.size
            for (j in pairs[i].second) {
                ans[j] = psum
            }
        }

        return ans


    }
}
